### PR TITLE
Add some tests for execution order for yield* in async generator

### PR DESCRIPTION
**Pull request by arai-a**
I'm about to write tests for yield* in async generator.
Here's tests for some basic flow, about execution order.
can I get some feedback before start writing more tests for abrupt cases?

### CODE REVIEW

File: test/language/statements/async-generator/yield-star-async-next.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    assert.sameValue(log[16].name, "after yield*");
+    assert.sameValue(log[16].value, "next-value-2");
+
+    assert.sameValue(v.value, "return-value");
+    assert.sameValue(v.done, true);
+
+    assert.sameValue(log.length, 17, "log.length");
+  }).then($DONE, $DONE);
+}).catch($DONE);
```
Comment (domenic) [inline test/language/statements/async-generator/yield-star-async-next.js]: Maybe any rejections should cause the test to fail?

File: test/language/statements/async-generator/yield-star-sync-next.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+              }
+            };
+          };
+        }
+      };
+    };
+  },
+  get [Symbol.asyncIterator]() {
+    log.push({ name: "get [Symbol.asyncIterator]" });
+    return null;
```
Comment (domenic) [inline test/language/statements/async-generator/yield-star-sync-next.js]: I'm probably missing something, but shouldn't this cause an error? Only undefined triggers the fallback to sync iterator, I thought?

Comment (caitp) [inline test/language/statements/async-generator/yield-star-sync-next.js]: The spec language agrees with you, this should be a TypeError trying to invoke `null`. V8 gets this wrong too, though. Maybe it would make sense to use `null` as a sentinel to use Async-from-Sync Iterator too?

Comment (domenic) [inline test/language/statements/async-generator/yield-star-sync-next.js]: Oh, we were reading the spec wrong. GetMethod censors `null` to `undefined`, so V8 and the test are right.

Comment (caitp) [inline test/language/statements/async-generator/yield-star-sync-next.js]: Ah, I see. Whew!

File: test/language/statements/async-generator/yield-star-async-next.js (hunk 1 above)
Comment (arai-a) [inline test/language/statements/async-generator/yield-star-async-next.js]: What does this comment mean?

Comment (domenic) [inline test/language/statements/async-generator/yield-star-async-next.js]: Meaning, right now, if the promise rejects, the test will just be marked as done. Maybe the test should fail. E.g. maybe `.catch(e => assert(false, "Must not reject"))` is more appropriate?

Comment (caitp) [inline test/language/statements/async-generator/yield-star-async-next.js]: `$DONE` should cause a test failure if the argument passed to it is truthy. (https://github.com/tc39/test262/issues/848)

Comment (domenic) [inline test/language/statements/async-generator/yield-star-async-next.js]: Ah OK, and I suppose it's probably safe to assume this won't somehow reject with undefined; seems good.

File: test/language/statements/async-generator/yield-star-async-next.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+              name: "call next",
+              args: args
+            });
+
+            nextCount++;
+            if (nextCount == 1) {
+              return {
+                get then() {
+                  log.push({ name: "get next then (1)" });
+                  return function(...args) {
```
Comment (caitp) [inline test/language/statements/async-generator/yield-star-async-next.js]: For the sake of readability, how about `return function(resolve) { .... }`

and assign args with `args: [...arguments]`,

and call resolve as `resolve` rather than as `args[0]`?

File: test/language/statements/async-generator/yield-star-async-next.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      });
+      var nextCount = 0;
+      return {
+        get next() {
+          log.push({ name: "get next" });
+          return function(...args) {
+            log.push({
+              name: "call next",
+              args: args
+            });
```
Comment (arai-a) [inline test/language/statements/async-generator/yield-star-async-next.js]: Do you think it's better checking thisValue for `next`, `return`, `throw`, and `then` ?
currently those functions' thisValue (object) are not global variable and cannot compare directly.
and putting them on global will make the test hard to follow, since the state (`nextCount` etc) needs to also be global.

would it be fine to define `name` property or something on those objects and compare `thisValue.name` against them?

Comment (arai-a) [inline test/language/statements/async-generator/yield-star-async-next.js]: maybe, I just need to put expected this value to `log`.

Comment (domenic) [inline test/language/statements/async-generator/yield-star-async-next.js]: I think the name solution is pretty good personally.

Comment (arai-a) [inline test/language/statements/async-generator/yield-star-async-next.js]: added thisValue and thisValue.name check for each getter/method

File: test/language/statements/async-generator/yield-star-async-next.js (hunk 3 above)
Comment (arai-a) [inline test/language/statements/async-generator/yield-star-async-next.js]: fixed.

**Comment by caitp** [comment]
Lgtm, thanks for these. Found a few bugs in the v8 impl!
